### ISSUE TITLE
Support for belongsTo through

### ISSUE THREAD

**Issue by Verdier** [sequelize/sequelize#3845]
Is there a plan for supporting `bellongsTo.through` such as:

``` javascript
A.hasMany(B)
A.hasMany(C) // Through B

B.belongsTo(A)
B.hasMany(D)

C.belongsTo(B);
C.belongsTo(A) // Through B
```

**Comment by janmeier** [comment]
Through is only usefull for n:m which is handled by belongsToMany

**Comment by Verdier** [comment]
Why no for N:1 such as in my previous exemple, I need this in a real world exemple:
```javascript
Operation.belongsTo(User)
Operation.hasMany(Breakdowns)

/* There is no way to define this relation without adding a 
field to the breakdown, but I will have to manually sync 
Operation.userId and Breakdon.userId */
Breakdowns.belongsTo(User) 
```

**Comment by janmeier** [comment]
If there is always only a single entry on one side of the relation, why do you need a join table?

Perhaps you can draw the relations, I can't really figure out your datamodel in my head :)

**Comment by mickhansen** [comment]
@Verdier You don't need a join table for 1:M/M:1. Your `Operations.hasMany(Breakdowns)` should work and add a foreign key to Breakdown refering to the operation.

**Comment by mickhansen** [comment]
Scratch my previous comment.

@Verdier You are looking to fetch a relation via another relation? Rails support this i believe, we don't yet i'm afraid.

**Comment by Verdier** [comment]
@mickhansen yep, the query would be something like that:
`...from breakdowns, operations, users where operations.id = breakdowns.id and users.id = operations.id`
No problem, I have solution for now.

**Comment by Verdier** [comment]
@janmeier the model is something like that:
```
User -> Operation -> Breakdown
```
A user have multiple operations, an operation have multiple breakdowns, but: a breakdown and an operation have only one same user. The `operation` works as a join table between one `breakdown` and one `user`. 

I now that I can store the `userId` in the `breakdown` AND in the `operation` but I will have to maintain the `userId` fields (ie `operation.userId === breakdown.userId` *always*).

**Comment by mickhansen** [comment]
Or you can use a nested include.

**Comment by Verdier** [comment]
@mickhansen yep, but that's mean that if you fetch the user's breakdown, you'll fetch, for each breakdown, the operation. That's not very efficient. Something like:
```javascript
{
 include: [{
 model: Operation,
 include: false, // Search for a condition in nested models, but do not eager load.
 where: {
  userId: userId
 }
}
```
would be very usefull.

**Comment by mickhansen** [comment]
@Verdier You'd need the join in any case.

**Comment by Verdier** [comment]
@mickhansen are you open for a pull request to support this feature:
```javascript
Breakdowns.belongsTo(User, {
 through: Operation
});
```

**Comment by mickhansen** [comment]
@Verdier You can disable attributes on the join with `include.attributes = []`.
Yes we'd be open to a pull request :)
The argument to `through` should maybe be an Association rather than a Model, for less ambiguity, not sure.

**Comment by janmeier** [comment]
Not sure through is even a good name here - it might be too confusing..

**Comment by mickhansen** [comment]
@janmeier Yeah i believe that's why Rails use through for this and pivot for our through.

**Comment by Verdier** [comment]
Here is a good article about `hasOne through` and `belongsTo through` using ActiveRecord: http://www.rojotek.com/blog/2014/05/16/cool-stuff-you-can-do-with-rails-has_one/.

**Comment by overlookmotel** [comment]
Related issue: https://github.com/sequelize/sequelize/issues/3935

**Comment by felixfbecker** [comment]
@mickhansen I just find the name `belongsToMany` so confusing. The difference between `hasMany` and `belongsToMany` is not clear at all.
n:m in other languages:
 - [Ruby](http://guides.rubyonrails.org/association_basics.html) calls n:m relations `has_and_belongs_to_many`, that makes it more clear that it is basically `hasMany`, but both ways.
 - [Doctrine](http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/association-mapping.html#many-to-many-bidirectional) (PHP) uses `@hasMany` on both models with `mappedBy` and `inversedBy` parameters. The join table is set through `@JoinTable`.
 - [Hibernate](https://docs.jboss.org/hibernate/orm/3.3/reference/en/html/associations.html#assoc-bidirectional-join-m2m) uses `many-to-many` columns on both models that both specify the same join table name `table`

I define my n:m relations in Sequelize in a similar way to the above:

**Project.ts**
```js
export const Project = sequelize.define('Project', {...});

import {User} from './User';
export const ProjectParticipants = Project.belongsToMany(User, {as: 'participants', through: 'UserProjects'});
```

**User.ts**
```ts
export const User = sequelize.define('User', {...});

import {Project} from './Project';
export const UserProjects = User.belongsToMany(Project, {as: 'projects', through: 'UserProjects'});
```

**Usage**
```ts
import {User, UserProjects} from './User';
const users = await User.find({include: [UserProjects]});
```
Basically I supply `through` as the join table name. And imo this would be nicer API if I could use `hasMany` for this.

**Comment by mickhansen** [comment]
@felixfbecker We actually have `hasAndBelongsToMany` as an alias for `belongsToMany`.
I understand your issue, but i don't think it's something we're likely to change right now, i prefer N:M being a seperate call from 1:M

**Comment by felixfbecker** [comment]
@mickhansen did not know that, thanks! And I also think it's better to keep it, just wanted to share my thoughts.

**Comment by overlookmotel** [comment]
If we think it's confusing, maybe make `hasAndBelongsToMany` the main name listed in the docs with `belongsToMany` as the alias, rather than the other way around?

**Comment by overlookmotel** [comment]
On the subject of `through`, I feel that `through` should refer to the *model* name rather than the *table* name.

Two reasons:

1. With every other interface in Sequelize, you use model names, so this stands out as an anomaly.
2. The over-arching principle of an ORM is to abstract away the database tables, so this breaks that principle.

A small point perhaps, but I feel consistency is really important to get this "it just works" effect.

**Comment by heisian** [comment]
Have this issue as well. As a workaround my thoughts are:

@Verdier OP's example:
* Give `Operations` a UNIQUE constraint on the `breakdown_id`
* Any `Breakdown` can only be associated once (to `Operation`), effectively giving the `Breakdown` only one possible parent.
* Add the helper for `Breakdown`:
```node
Breakdown.prototype.getUser = async function() {
  const [user] = await this.getUsers()
  return user
}
```

@felixfbecker's example:
* Give `UserProjects` a UNIQUE constraint on the `project_id`
* Any `Project` can only be associated once (to `User`), effectively giving the `Breakdown` only one possible parent.
* Add the helper for `Project`:
```node
Project.prototype.getUser = async function() {
  const [user] = await this.getUsers()
  return user
}
```

And finally, for my glorious example:
```
Referrers => Referral Programs => Referrals
```
A Referrer can have any number of Referral Programs that can have any number of Referrals.
A Referral can only belong to one Referral Program which can only belong to one Referrer.
A Referral belongs to one Referrer.

**Comment by hassangit** [comment]
@Verdier OK this magically ✨ work's for me, @mickhansen @janmeier could any of you please explain why maybe it's already implemented there and useful I've seen a lot of requests for that feature.
even though the option `through` is not possible in `belongsTo` I've tried it and what makes it work that I've passed also `foreignKey: 'id'` which refers to the `A` FK in the join table `C`  
```
// A --> B --> C

C.belongsTo(A, {
  through: B,
  foreignKey: 'id',
})
```
if this has no unexpected behavior, the next step is how do we assign even more far nested joins like it
.ie  
```
// A --> B --> C --> D

D.belongsTo(A, { 
  ... how ?
})
```
thank you ❤

**Comment by zocoi** [comment]
@hassangit your example doesn't work for me, what do you do exactly?

**Comment by odravison** [comment]
I passed by a lot of problem because Sequelize doesn't support this feature. Would be very nice to have this on this project!
It's very useful.

**Comment by simonbrent** [comment]
I have noticed that if you do 
```
var association = modelA.belongsToMany(modelB, { through: 'model_c', ... });
association.isMultiAssociation  = false;
association.isSingleAssociation = true;
```
then you get back the associated data as a single item rather than an array, while the SQL query does not change.

**Comment by david-mart** [comment]
> I have noticed that if you do
> 
> ```
> var association = modelA.belongsToMany(modelB, { through: 'model_c', ... });
> association.isMultiAssociation  = false;
> association.isSingleAssociation = true;
> ```
> 
> then you get back the associated data as a single item rather than an array, while the SQL query does not change.

this actually works, thank you for a tip!

**Comment by rdsedmundo** [comment]
@ephys @WikiRik do you think this could be re-opened? I believe it's still relevant and the stale bot closed it a while ago.

I have another use case for this, which adds to what OP requested. I'm using an association scope and through an unique key I know that there's one only row that will satisfy it through the junction table, but I'd like to receive a single item instead of an array. 

```
   // #1 scenario: a User has many Groups
    models.User.belongsToMany(models.Group, {
      foreignKey: 'userId',
      through: UserGroups,
    });

   // #2 scenario: but a User only owns a single personal group
    models.User.belongsToMany(models.Group, {
      as: 'primaryGroup',
      foreignKey: 'userId',
      scope: { 
        groupType: 'personal' 
      },
      through: {
        model: UserGroups,
        scope: {
          isGroupOwner: true,
        },
      },
    });
```

I know I could probably just set that the User owns the Group elsewhere, but it's desirable to have it centralized it in the junction UserGroup table, as it's subject to change. 

The workaround provided by @simonbrent works, but it feels a bit hacky and I suppose this could change in future Sequelize versions since it seems to be overriding internal variables.

This could also potentially come as an option in `belongsToMany` like `isSingleRow` or whatever.
